---
title: "MR-Hevo"
author: Paul McKeigue
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{MR-Hevo}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r opts, include = FALSE, eval = TRUE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
  )

# devtools::check(vignettes=FALSE)
# devtools::document()
# devtools::build(vignettes=FALSE)
# R CMD INSTALL --library=~/R/x86_64-pc-linux-gnu-library/3.6 ./mrhevo_0.0.0.9000.tar.gz

```

## Introduction

Conventional 'estimators' for Mendelian randomization deal with pleiotropy by downweighting or excluding those instruments that appear to be outliers.  As explained in the README document and in a [description of the statistical model](https://github.com/molepi-precmed/mrhevo/blob/main/theorymethods.pdf), MR-Hevo tries to learn the distribution of pleiotropic effects over loci, and marginalizes over this distribution to compute the likelihood as a function of the causal effect parameter. 

## Input dataset

The input dataset should comprise: 

* `Y` a vector of length `N` containing outcomes (in this example type 1 diabetes as a binary outcome

* `Z` an `N x J` data.table of genetic instruments

* `X_u` an `N x U` data.table of unpenalized covariates (this will usually include scores on the first few principal components of the genotype matrix

* `Z.metadata` a data.table with `J` rows, one for each genetic instrument.  Estimated coefficients for the effects of instruments `Z` on the exposure, and the corresponding standard errors of these estimates are in columns named `alpha_hat` and `se_alpha_hat` respectively.  The column `scoreid` should contain the names of the columns of `Z`. This table can include other metadata such as the genomic position of each genotypic instrument.  

In this example the genetic instruments consist of 5 _trans_-eQTLs for the expression of the gene _CTLA4_ in whole blood.  The eQTLs are far enough apart to be modelled as independent, and the outcome is type 1 diabetes.  The unpenalized covariates are sex and the first five principal components of the genotype matrix.  There are 2124 cases and 10620 controls.  

Using MR-Hevo with only 5 instruments is **not** recommended in practice, but it's useful for this tutorial to work with an example where inference is only just possible.  To learn the distribution of pleiotropic effects over instruments, the model-fitting algorithm needs as many instruments as possible. 



```{r setup, eval=TRUE}

library(data.table)
library(ggplot2)
library(rstan)
require(knitr)

source("functions.mrhevo.R")

load("data_mrhevo.RData")

if(length(unique(Y)) == 2) {
  logistic <- TRUE
} else {
  logistic <- FALSE
}

```

As a first step, we compute summary statistics rom which we obtain an estimate of the causal effect from each "instrument"

## Summary statistics

```{r summarystats, include=TRUE, eval=TRUE}

coeffs.scores <- get_summarystatsforMR(Y, Z, X_u)

coeffs.scores <- Z.metadata[coeffs.scores, on="scoreid"]
coeffs.scores[, thetaIV := beta_YZ / alpha_hat]  # ratio estimates

knitr::kable(coeffs.scores[, .(scoreid, qtl_type, alpha_hat, se.alpha_hat,
                               beta_YZ, SE_YZ, thetaIV, z, p)])
```

A plot of `beta_YZ` against `alpha_hat` shows that one of the instruments is an outlier, with a much larger ratio of `beta_YZ` to `alpha_hat` than the others.  The effect of the _cis_-eQTL is in the opposite direction to the effect of the _trans_-eQTLs: this is explicable, as the _cis_- acting SNPs alter the splicing of _CTLA4_.  


```{r plotcoeffs, include=TRUE, eval=TRUE, fig.width=5}

p.coeffs <- ggplot(coeffs.scores, aes(x=alpha_hat, y=beta_YZ,
                                                        color=qtl_type)) +
    geom_point() +
    scale_x_continuous(limits = c(0, NA))
p.coeffs

```

From the summary statistics (excluding the _cis_-eQTL) we can obtain three widely-used estimators of the causal effect parameter: the inverse-variance weighted mean, the weighted median, and the penalized weighted median.  As expected, the weighted median and the penalized weighted median downweight the outlier and give less extreme estimates of the causal effect.

## Estimators of the causal effect based on summary statistics

```{r estimators, eval=TRUE}

estimators.dt <- get_estimatorsMR(coeffs.scores[qtl_type=="trans"])[, 1:5] 

knitr::kable(estimators.dt)

```

## Sample the posterior density

The function `run_mrhevo` will call `Stan` to sample the posterior distribution of all parameters, given the model and the data.  It returns an object of class `stanfit`

As recommended by Piironen and Vehtari (2017), we supply a prior guess of the fraction of instruments that have nonzero pleiotropic effects.  With a large enough sample size, this prior guess will not make much difference to the results. For this example, we guess the fraction to be 0.2.  

The causal effect parameter `theta` is assigned a Gaussian prior with mean zero and standard deviation `priorsd.theta`.  As we will divide the posterior by the prior to obtain the likelihood of `theta`, this prior should not make any difference but the sampler will be more efficient if the prior standard deviation is set to a value that is supported by the data. 

Sampling 4 chains parallelized over 4 cores takes about 20 minutes with this dataset of about 12000 observations.  Setting the argument `use.sampling` to `FALSE` will run `Stan` with a variational approximation to the posterior distribution, which is much faster.  Experience so far is that this approximation does not work well with MR-Hevo, and should be used only to test the analysis pipeline. 

```{r sampling, eval=TRUE}

options(mc.cores = parallel::detectCores())
rstan_options(auto_write = TRUE)

use.sampling <- TRUE
logistic <- TRUE
priorsd.theta <- 1
newrun <- FALSE

which.trans <- colnames(Z) %in% Z.metadata[qtl_type=="trans", scoreid]

if(newrun) {
    hevo.stanfit <- run_mrhevo(use.sampling=use.sampling, logistic=logistic,
                               Z[, ..which.trans], Y,
                               sigma_y=NULL, X_u,
                               alpha_hat=Z.metadata[qtl_type=="trans", alpha_hat],
                               se.alpha_hat=Z.metadata[qtl_type=="trans", se.alpha_hat],
                     fraction_pleio=0.2, priorsd_theta=priorsd.theta)
    save(hevo.stanfit, file="hevo.stanfit.RData")
} else {
   load("hevo.stanfit.RData")
}

```

## Maximum likelihood estimate and _p_-value
From the posterior samples, we compute the likelihood as a function of `theta`.  We fit a quadratic approximation to the log-likelihood, and use this to obtain a classical maximum likelihood and _p_-value.  

A plot of the posterior density of `theta` shows that it is approximately Gaussian.  A corollary of this is that the log-likelihood can be approximated by a quadratic function, and that the value of `theta` that maximizes the likelihood will have the sampling properties -- consistency, unbiasedness, and minimum variance -- of a good "estimator". 

```{r mle, eval=TRUE, fig.width=5, fig.asp=1.2}

theta.samples <-  unlist(extract(hevo.stanfit, pars="theta"))
prior.theta <- dnorm(theta.samples, mean=0, sd=priorsd_theta)
mle.theta <- mle.se.pval(x=theta.samples, prior=prior.theta)
mle.theta[, method := "Marginalize over direct effects"]

p.bayesianloglik <-  mle.se.pval(x=theta.samples, prior=prior.theta, return.asplot=TRUE)
p.bayesianloglik

```

## Comparison with other estimates

Marginalizing over the distribution of unobserved pleiotropic effects is more conservative than the standard methods for inferring the causal effect parameter, at least in this example.  With only 5 instruments, we don't expect to be able to learn much about the distribution of pleiotropic effects over instruments. 


```{r methodscomparison, eval=TRUE}

methods.dt <- rbind(estimators.dt, mle.theta, fill=TRUE)
knitr::kable(methods.dt[, .(method, Estimate, SE, z, pvalue)])

```

## Shrinkage coefficients
For each instrument, we can compute from the model parameters a shrinkage coefficent `kappa` that takes values between 0 (no shrinkage of the pleiotropic effect) and 1 (complete shrinkage of the pleiotropic effect to zero).  The posterior distribution of the `J` shrinkage coefficients has the horseshoe shape specified by the prior. 


```{r kappa, eval=TRUE}

kappa.all <- extract(hevo.stanfit, pars="kappa")$kappa
kappa.dt <- data.table(kappa=as.numeric(kappa.all))

p.kappa <- ggplot(kappa.dt, aes(x=kappa)) +
  geom_histogram(aes(y=after_stat(density)),
                 breaks=seq(0, 1, by=0.05), color="black", fill="gray") +
  labs(x = expression(paste("Shrinkage coefficient ", kappa, " imposed on direct effects of instruments")),
       y= "Posterior density")
p.kappa

```

## Coefficients for pleiotropic effects
Summaries of the posterior distribution show that the pleiotropic effect `beta[1]` of instrument 1 has escaped shrinkage, but the pleiotropic effects of the other four instruments have been shrunk to zero.  

```{r summary.params, eval=TRUE}

if(use.sampling) {
  pars.forsummary <- c("theta", "beta", "log_c", "log_tau", "m_eff", "lp__")
} else {
  pars.forsummary <- c("theta", "beta", "log_c", "log_tau", "m_eff")
}

fit.coeffs <- summary(hevo.stanfit,
                      pars=pars.forsummary, 
                      probs=c(0.1, 0.9))$summary 
fit.coeffs <- as.data.table(round(fit.coeffs, 2), keep.rownames="variable")

knitr::kable(fit.coeffs)

```

## Sampler diagnostics
A pairs plot is useful to diagnose problems with the sampler

```{r diagnostics, eval=TRUE, fig.width=6, fig.asp=1}

if(use.sampling) {
  pairs(hevo.stanfit, pars=c("theta", 
                     "log_c", "log_tau", "m_eff",
                     "lp__"))
}

```


